Linux: Initial support for Gentoo installations.

rkitover opened the pull request:
Add Gentoo to list of supported distributions and add a
tools/installpsh-gentoo.sh script which installs the Ubuntu 18.04
powershell or powershell-preview .deb using the Gentoo dpkg port.

If -includeide is specified, also installs Visual Studio Code via
flatpak, installing flatpak-overlay and flatpak if necessary.

This should work for most Gentoo users with a relatively recent system.

Also fix sudo check and pwsh-preview test in tools/installpsh-debian.sh.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=11429) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=11429) before we can accept your contribution.<br/><br/>:x: rkitover [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=11429)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=11429) it.</sub>

rkitover commented:
Fixed removing the temp directory and made the output nicer.

rkitover commented:
Added support for `-includeide` via flatpak.

rkitover commented:
Made more improvements, it now finds the latest ubuntu dist automatically and doesn't use `/tmp`.

rkitover commented:
I am pretty much done with this. I know it's the holidays and everything, but can someone take a look please? I can work on other stuff separately.

rkitover committed:
Linux: Initial support for Gentoo installations.

Add Gentoo to list of supported distributions and add a
tools/installpsh-gentoo.sh script which installs the latest Ubuntu
powershell or powershell-preview .deb using the Gentoo dpkg port.

If -includeide is specified, also installs Visual Studio Code via
flatpak, installing flatpak-overlay and flatpak if necessary.

This should work for most Gentoo users with a relatively recent system.

Also fix sudo check and pwsh-preview test in tools/installpsh-debian.sh.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

daxian-dbw commented:
@anmenaga Could you please review this PR? Thanks!

rkitover commented:
@daxian-dbw I realize I made some opinionated decisions here, but horrible hacks are often the best solution on gentoo. I might work on a proper ebuild in the main gentoo repository for this, but I think this solution is not terrible for the time being and will work for most gentoo users.

Also this code does not affect the other installers much, so there is little potential for breakage.

I am also happy to contribute to these scripts in the future.

Let me know if you need anything else.

daxian-dbw commented:
@rkitover My apology for the delay. I will drive the review process on our side. Thank you for your patience!

daxian-dbw commented:
@anmenaga @TravisEz13 Can you please review this PR after the 7.0.0-rc.3 release? Thank you!

daxian-dbw commented:
@anmenaga @TravisEz13 A gentle ping :) Please take a look at this PR, thanks!

anmenaga reviewed (commented):
Looks mostly fine to me except that it pulls some stuff from 3rd parties:
1) `https://github.com/fosero/...`
2) `https://raw.githubusercontent.com/DarwinJS/...`

@TravisEz13  can you please look at this? Thank you.

rkitover commented:
Thank you very much for looking at this, it's not very important of course, but I might do more work on these scripts too.

Just a note that the DarwinJS URL is taken straight from the debian script for the vscode extension.

rkitover commented:
@daxian-dbw ping,

I have one review.

Explanation of URLs:

| **URL**                                                           | **Description**                                                            |
|-------------------------------------------------|-----------------------------------------------------------|
| https://github.com/fosero/flatpak-overlay.git | This is a third party repository for gentoo, it has the flatpak ebuild which is not in core gentoo. |
| https://raw.githubusercontent.com/DarwinJS/CloudyWindowsAutomationCode/master/pshcoredevenv/testpowershell.ps1 | This is taken straight from `installpsh-debian.sh`, it's a script to test the Visual Studio Code PowerShell plugin installation. |

I'd like to get more involved in this project, I have been waiting on this PR.

Thank you!

rkitover commented on tools/installpsh-gentoo.sh:
```diff
@@ -0,0 +1,220 @@
+
+#gitrepo paths are overrideable to run from your own fork or branch for testing or private distribution
+
+VERSION="1.2.0"
```
I will look into this, I've never worked with azure but I've done work with docker containers etc.. And it looks like some base images for gentoo are available.
